src/gui/guiFormSpecMenu.cpp

```diff
@@ -4228,153 +4286,384 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
 		u32 craft_amount = 0;
 
 		switch (updown) {
-		case BET_DOWN:
+		case BET_DOWN: {
 			// Some mouse button has been pressed
 
-			//infostream << "Mouse button " << button << " pressed at p=("
-			//	<< event.MouseInput.X << "," << event.MouseInput.Y << ")"
-			//	<< std::endl;
-
-			m_selected_dragging = false;
+			if (m_held_button >= 0)
+				break;
+			if (button == BET_LEFT || button == BET_RIGHT || button == BET_MIDDLE)
+				m_held_button = button;
 
-			if (s.isValid() && s.listname == "craftpreview") {
+			if (!s.isValid()) {
+				if (m_selected_item && !getAbsoluteClippingRect().isPointInside(m_pointer)) {
+					// Clicked outside of the window: drop
+					if (button == BET_RIGHT || button == BET_WHEEL_UP)
+						drop_amount = 1;
+					else if (button == BET_MIDDLE)
+						drop_amount = MYMIN(m_selected_amount, 10);
+					else if (button == BET_LEFT)
+						drop_amount = m_selected_amount;
+				}
+				break;
+			}
+			if (s.listname == "craftpreview") {
 				// Craft preview has been clicked: craft
-				craft_amount = (button == BET_MIDDLE ? 10 : 1);
-			} else if (!m_selected_item) {
-				if (s_count && button != BET_WHEEL_UP) {
-					// Non-empty stack has been clicked: select or shift-move it
+				if (button == BET_MIDDLE)
+					craft_amount = 10;
+				else if (event.MouseInput.Shift && button == BET_LEFT)
+					// TODO: We should craft everything with shift-left-click,
+					// but the slow crafting code limits us, so we only craft one
+					craft_amount = 1;
+					//craft_amount = list_s->getItem(s.i).getStackMax(m_client->idef());
+				else
+					craft_amount = 1;
+
+				// Holding shift moves the crafted item to the inventory
+				m_shift_craft = event.MouseInput.Shift;
+
+			} else if (!m_selected_item && button != BET_WHEEL_UP && !empty) {
+				// Non-empty stack has been clicked: select or shift-move it
+				u32 count;
```

Comment:
This variable is possibly uninitialized. Whereas it might be appropriately protected, this will not always be the case in the future.


```suggestion
				u32 count = 0;
```